src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -0,0 +1,1125 @@
+                                found_request.request_ids.correlation_id &&
+                            request.request_ids.function_id ==
+                                found_request.request_ids.function_id;
+                   });
+  list_of_sent_requests.erase(request_iter);
+
+  if (!list_of_sent_requests.empty()) {
+    LOG4CXX_DEBUG(logger_,
+                  "Resumption app "
+                      << app_id << " not finished . Amount of requests left : "
+                      << list_of_sent_requests.size());
+    return false;
+  }
+  return true;
+}
+
+utils::Optional<ResumeCtrl::ResumptionCallBack>
+ResumptionDataProcessor::GetResumptionCallback(const uint32_t app_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  sync_primitives::AutoReadLock lock(register_callbacks_lock_);
+  auto it = register_callbacks_.find(app_id);
+  if (it == register_callbacks_.end()) {
+    LOG4CXX_WARN(logger_, "Callback for app_id: " << app_id << " not found");
+    return utils::Optional<
+        ResumeCtrl::ResumptionCallBack>::OptionalEmpty::EMPTY;
+  }
+  return utils::Optional<ResumeCtrl::ResumptionCallBack>(it->second);
+}
+
+bool ResumptionDataProcessor::IsResumptionSuccessful(const uint32_t app_id) {
+  sync_primitives::AutoWriteLock lock(resumption_status_lock_);
+  ApplicationResumptionStatus& status = resumption_status_[app_id];
+  return status.error_requests.empty() &&
+         status.unsuccessful_vehicle_data_subscriptions_.empty();
+}
+
+void ResumptionDataProcessor::EraseAppResumptionData(
+    const uint32_t app_id,
+    const hmi_apis::FunctionID::eType function_id,
+    const int32_t corr_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  resumption_status_lock_.AcquireForWriting();
+  resumption_status_.erase(app_id);
+  resumption_status_lock_.Release();
+
+  request_app_ids_lock_.AcquireForWriting();
+  request_app_ids_.erase({function_id, corr_id});
+  request_app_ids_lock_.Release();
+
+  register_callbacks_lock_.AcquireForWriting();
+  register_callbacks_.erase(app_id);
+  register_callbacks_lock_.Release();
+}
+
+void ResumptionDataProcessor::ProcessResponseFromHMI(
+    const smart_objects::SmartObject& response,
+    const hmi_apis::FunctionID::eType function_id,
+    const int32_t corr_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_TRACE(logger_,
+                "Now processing event with function id: "
+                    << function_id << " correlation id: " << corr_id);
+
+  auto found_app_id = GetAppIdWaitingForResponse(function_id, corr_id);
+  if (!found_app_id) {
+    return;
+  }
+  const uint32_t app_id = *found_app_id;
+  LOG4CXX_DEBUG(logger_, "app_id is: " << app_id);
+
+  auto found_request = GetRequest(app_id, function_id, corr_id);
+  if (!found_request) {
+    return;
+  }
+  auto request = *found_request;
+
+  ProcessResumptionStatus(app_id, response, request);
+
+  if (!IsResumptionFinished(app_id, request)) {
+    return;
+  }
+
+  auto found_callback = GetResumptionCallback(app_id);
+  if (!found_callback) {
```

Comment:
ERROR logging would be useful

Reply:
Done in eb410750791e0a6b8d8d4090b1ce53f1d547c0fd